activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -639,18 +639,37 @@ def default_index_type?(index) # :nodoc:
       end
 
       def build_insert_sql(insert) # :nodoc:
-        sql = +"INSERT #{insert.into} #{insert.values_list}"
-
-        if insert.skip_duplicates?
-          no_op_column = quote_column_name(insert.keys.first)
-          sql << " ON DUPLICATE KEY UPDATE #{no_op_column}=#{no_op_column}"
-        elsif insert.update_duplicates?
-          sql << " ON DUPLICATE KEY UPDATE "
-          if insert.raw_update_sql?
-            sql << insert.raw_update_sql
-          else
-            sql << insert.touch_model_timestamps_unless { |column| "#{column}<=>VALUES(#{column})" }
-            sql << insert.updatable_columns.map { |column| "#{column}=VALUES(#{column})" }.join(",")
+        no_op_column = quote_column_name(insert.keys.first)
+
+        # Avoid MySQL 8.0 deprecation warning, see https://dev.mysql.com/worklog/task/?id=13325.
+        if !mariadb? && database_version >= "8.0.0"
+          values_alias = quote_table_name("#{insert.model.table_name}_values")
+          sql = +"INSERT #{insert.into} #{insert.values_list} AS #{values_alias}"
+
+          if insert.skip_duplicates?
+            sql << " ON DUPLICATE KEY UPDATE #{no_op_column}=#{values_alias}.#{no_op_column}"
+          elsif insert.update_duplicates?
+            sql << " ON DUPLICATE KEY UPDATE "
+            if insert.raw_update_sql?
+              sql << insert.raw_update_sql
```

Comment:
> Note: we have been ignoring the MySQL warning

Currently, we have a couple of places that provide a `raw_update_sql` pointing to the `VALUES(...)`, which, with this change, created a mixed query:

```
INSERT table (...) VALUES (...) as table_alias ON DUPLICATE KEY UPDATE x=VALUES(x)
```

The query didn't work as expected, and the expected behavior broke. We may need some way to keep using the previous behavior for a while. (I'm working on a PR unless someone beat me)

cc: @matthewd

Reply:
Yes. If the `raw_update_sql` is provided, we should not try to alias, because we don't even know the alias.

Reply:
Created this [PR](https://github.com/rails/rails/pull/51325) to address this issue.